services/core/IEEE_2030_5/ieee_2030_5/client.py

```diff
@@ -0,0 +1,649 @@
+                self._update_list(ed.FunctionSetAssignmentsListLink.href, 
+                                  "FunctionSetAssignments", self._fsa_map, self._fsa)
+                
+                if ed.DERListLink:
+                    derlist: m.DERList = self.__get_request__(ed.DERListLink.href)
+                    self._der_map[derlist.href] = derlist
+                    for index, der in enumerate(derlist.DER):
+                        self._der[der.href] = der
+            
+            for fsa in self._fsa.values():
+                if fsa.DERProgramListLink:
+                    self._der_program_map[fsa.DERProgramListLink.href] = self.__get_request__(fsa.DERProgramListLink.href)
+                    
+            
+                            
+        if dcap.MirrorUsagePointListLink is not None and dcap.MirrorUsagePointListLink.href:
+            self._update_list(dcap.MirrorUsagePointListLink.href, "MirrorUsagePoint", self._mirror_usage_point_map, self._mirror_usage_point)
+        
+        
+        # if dcap.UsagePointListLink is not None and dcap.UsagePointListLink.href:
+        #     self._update_list(dcap.UsagePointListLink.href, "UsagePoint", self._usage_point_map, self._usage_point)
+        
+        self._dcap = dcap 
+        
+        
+    def post_log_event(self, end_device: m.EndDevice, log_event: m.LogEvent):
+        if not log_event.createdDateTime:
+            log_event.createdDateTime = self.server_time
+        
+        self.request(end_device.LogEventListLink.href, method="POST")    
+        
+                
+                
+    def _update_list(self, path: str, list_prop: str, outer_map: Dict, inner_map: Dict):
+        """Update mappings using 2030.5 list nomoclature.
+        
+        Example structure for EndDeviceListLink
+        
+            EndDeviceListLink.href points to EndDeviceList.
+            EndDeviceList.EndDevice points to a list of EndDevice objects.
+            
+        Args:
+        
+            path: Original path of the list (in example EndDeviceListLink.href)
+            list_prop: The property on the object that holds a list of elements (in example EndDevice)
+            outer_mapping: Mapping where the original list object is stored by href
+            inner_mapping: Mapping where the inner objects are stored by href
+        
+        """
+        my_response = self.__get_request__(path)
+        
+        if not self._is_ok:
+            raise RuntimeError(my_response)
+        
+        if my_response is not None:
+            href = getattr(my_response, "href")
+            outer_map[href] = my_response
+            for inner in getattr(my_response, list_prop):
+                href = getattr(inner, "href")
+                inner_map[href] = inner
+            
+        
+    def _get_device_capabilities(self, endpoint: str) -> m.DeviceCapability:
+        dcap: m.DeviceCapability = self.__get_request__(endpoint)
+        if self._response_status != 200:
```

Comment:
some parts use _is_ok and and some check for response code. This particular place also assumes response is set in self._response_status. Better design might be for get/put methods to return response and _is_ok to accept that as param and return True/False